prov/gni: scalable memory registration

jswaro opened the pull request:
This pull request implements a functional version of scalable memory registration. Provider memory registration optimizations have not yet been implemented.

closes #458 #1380 #1381 #1378 #1217 #1188 #1244 

@sungeunchoi @hppritcha @chuckfossen

jshimek commented on man/fi_gni.7.md:
```diff
@@ -135,18 +152,49 @@ struct fi_gni_ops_domain {
-                             uint64_t key, enum fi_datatype datatype,
-                             enum gnix_fab_req_type req_type,
-                             void *context);
+		size_t (*native_amo)(struct fid_ep *ep, const void *buf,
```
Did you mean to muck with the whitespace here?

jshimek commented on prov/gni/include/gnix_bitmap.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015 Cray Inc. All rights reserved.
+ * Copyright (c) 2015-2017 Cray Inc. All rights reserved.
```
If we didn't modify this file last year we shouldn't be using the -, it should probably just be Copyright 2015, 2017.

hppritcha commented:
I think I'm going to need some more info about what's going on with the shared memory file and the Aries vMDH usage to review this PR effectively.

jswaro commented:
The simple explanation is that the shared memory file is a node local memory storage for vmdh indices. This allows processes running on the same node to know what index values are not in use, which is especially important for provider registrations, and also provides feedback to the user application if they request a key that is already in use by another process.

hppritcha commented:
I see.  Expect problems with that approach - if folks try to use it.  For example, for OpenSHMEM implementations, they will expect with the Mr. Scalable approach, that each process (PE) on a node can use a user specified key for the symmetric heap.  They're not going to expect that each PE will need to use a different key value.  I'm fine with reviewing for code correctness though.

jswaro commented:
> I see. Expect problems with that approach - if folks try to use it. For example, for OpenSHMEM implementations, they will expect with the Mr. Scalable approach, that each process (PE) on a node can use a user specified key for the symmetric heap.

That problem is outside of our hands. That is a hardware limitation for VMDH indices. On node, VMDH indices are unique on a per-ptag basis. If multiple processes use the same PTAG and VMDH (scalable), they will need to account for that with the GNI provider. 

The alternative, is to embed a larger amount of complexity to map vmdh indices to specific endpoints, which would require us to know the logical offset of the rank on node within the VMDH table. IE, if there are 8 processes on node using scalable with the same ptag, key=0 for rank 0 is 0, but key=0 for rank 1 is actually 16. (16 key stride).

hppritcha commented:
doesn't look like the LANL CI is happy with this PR.  Please fix the compile problem.

hppritcha commented on prov/gni/include/gnix_bitmap.h:
```diff
@@ -176,4 +189,12 @@ int _gnix_bitmap_full(gnix_bitmap_t *bitmap);
+/* helper function for determining the size of array needed to support
+   'x' number of bits for an externally provided buffer address */
+__attribute__((unused))
+static inline uint32_t _gnix_bitmap_get_buffer_size(int nbits)
```
a doxygen style comment might be nice here.

hppritcha commented on prov/gni/include/gnix_mr.h:
```diff
@@ -191,6 +195,16 @@ int _gnix_close_cache(struct gnix_fid_domain *domain,
+/* used for internal registrations,
+   Set reserved to 0 for a user registration,
+   Set reserved to 1 for a provider registration */
+int _gnix_mr_reg(struct fid *fid, const void *buf, size_t len,
```
a doxygen style comment here too would be good

hppritcha commented on prov/gni/src/gnix_auth_key.c:
```diff
@@ -36,13 +36,87 @@
+#define MAX_VMDH_REGS 4096
+#define BITMAP_ARR_SIZE(nbits) \
+	(GNIX_BITMAP_BLOCKS(nbits) * sizeof(gnix_bitmap_block_t))
+#define GNIX_VMDH_PATH "/tmp/gnix_vmdh_info"
```
How is this suppose to work if there are multiple jobs on a node, i.e. slurm with non exclusive mode?

hppritcha replied:
I also don't think its a good idea to hardware a filename that is opened in /tmp.  At least something that honors $TMPDIR environment variable.

jswaro replied:
> How is this suppose to work if there are multiple jobs on a node, i.e. slurm with non exclusive mode?

If you don't have multiple jobs sharing the same file, there will be issues with reporting useful feedback to the application. I need to modify the permissions to allow all users to edit the file, because otherwise it won't work with the MAMU case

> I also don't think its a good idea to hardware a filename that is opened in /tmp. At least something that honors $TMPDIR environment variable.

I can add an environment variable, but I'll add the cautionary warning here: If the environment is propagated uniformly, there could be issues -- However, that isn't our issue to resolve, so it's a minor thing.

hppritcha commented on prov/gni/src/gnix_shmem.c:
```diff
@@ -0,0 +1,182 @@
+				sleep(1);
+				wait_count++;
+			}
+		} while ((stat_info.st_size < size) &&
```
It would be nice to have a tunable here for how long to wait for the file to be ready.  I'm thinking of KNL nodes with very slow startup of processes under slurm.  check with MPT group for issues in that area.

jswaro replied:
Sure, I'll add that.

hppritcha commented on prov/gni/src/gnix_auth_key.c:
```diff
@@ -36,13 +36,87 @@
+	int ret;
+
+	if (!__gnix_vmdh_info) {
+		ret = _gnix_shmem_create(path,
```
How is access to this file serialized across separate processes?

jswaro replied:
If the file does exist, it is opened with exclusive access. The opener will create the memory structure and exit. Other files who attempt to open the file while in exclusive access by the owner, will wait until the file is the expected size, then attempt to open it after that condition is satisfied. 

Inside the memory segment, there is a lock for serializing access to critical sections. 

See gnix_shmem.c:60 , gnix_shmem.c:138 , gnix_auth_key.c:91, gnix_auth_key.c:168

jswaro commented:
> I'd like at least the name of the shared tmp file to be better.

I picked something and moved forward to continue development of the feature. Feel free propose a new one.

hppritcha commented:
For the file name, you could post-pend the job container id or something like that to prevent name collisions.

jswaro commented:
> For the file name, you could post-pend the job container id or something like that to prevent name collisions.

As I mentioned above, it would be best if jobs on node shared the file so that informative return codes and debug could be returned to the user. 

If the files are unique on a per-job basis, and multiple jobs run on the node, libfabric will have limited or no knowledge of which keys are being used. Given we have a limit of 4 PTAGs using VMDH, this also means less knowledge about how many PTAGs are using VMDH, and less informative messages if and when an application attempts to use more authorization keys with VMDH than we can support.

chuckfossen commented on man/fi_gni.7.md:
```diff
@@ -277,6 +325,36 @@ address format. In order to populate the remote peer's address vector
+
+All memory registrations are associated with an authorization key, whether it is the provider
+default key(keylen=0) or a user-acquired key (key!=NULL, keylen!=0). Each authorization
+key is associated with a unique GNI network key. A GNI network key can only accommodate a single
```
Let's add a space between key and (keylen=0)

chuckfossen commented on prov/gni/include/gnix_bitmap.h:
```diff
@@ -98,13 +108,16 @@ int _gnix_test_and_clear_bit(gnix_bitmap_t *bitmap, uint32_t index);
  * @param   bitmap  a gnix_bitmap pointer to the bitmap struct
  * @param   nbits   number of bits to request space for
+ * @param   addr    if provided, external memory allocation used for internal
+					array
```
Extra indentation here. We probably want spaces to match the surrounding code.

jswaro replied:
ok

chuckfossen commented on prov/gni/src/gnix_shmem.c:
```diff
@@ -0,0 +1,182 @@
+
+		/* Unexpected error? */
+		if (errno != EEXIST) {
+			saved_errno = errno;
```
What is the purpose of saved_errno?

jswaro replied:
I've found in other contexts with this type of approach, that if you have multiple failed system calls that can modify errno, that preserving the errno for the original error aids in debugging the actual problem.

```c
sz = failing_function();
if (sz == -1)
    /* let's assume that errno was something like enodev... 
        now to do something to report the error, but let's also 
         assume that it will also fail... */

   WARN(error_message, errno); 
   /* now depending on how WARN is implemented, it could 
       copy the reference to the stack at the correct time, or 
       it could make another failing call which modifies errno 
       prior to the actual printf, or other function. Because we 
       don't use a function with GNIX_WARN, it can't be safely 
       assumed that the value of errno will be safely saved on 
       the stack by the time the printf is done. In our debug 
       code, we perform sprintfs to get the node id, as well 
       as other things. This is only done to ensure everything 
       is saved safely. */

       // assume warn failed, and changed errno to ENOSPC, it may print the wrong message.

    /* Long story short: Preserving the errno here saves us 
        from implementation specific details of GNIX_WARN, 
        or other calls we might use here */
```

jswaro replied:
I'll fix this so it returns -saved_errno

chuckfossen commented on prov/gni/src/gnix_mr.c:
```diff
@@ -310,23 +385,121 @@ DIRECT_FN int gnix_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
+		addr = (uint64_t) modify->attr.mr_iov[i].iov_base;
+		len = (uint64_t) modify->attr.mr_iov[i].iov_len;
+
+		aligned_addr = addr & ~0xfff;
```
Let's use macros for hard coded values.

jswaro replied:
Will do. Forgot about that one.

chuckfossen commented on prov/gni/test/allocator.c:
```diff
@@ -41,6 +41,8 @@
 #include "gnix_rdma_headers.h"
 #include "common.h"
 
+#include "common.h"
```
Looks like common.h is included twice now or is the diff funky?

jswaro replied:
It was included twice. I'll fix it.

chuckfossen commented on prov/gni/test/api.c:
```diff
@@ -762,9 +999,11 @@ void do_atomic_write_fetch(void)
 	sz = fi_atomic(ep[0], source, 1,
-		       loc_mr[0], gni_addr[1], (uint64_t)target, mr_key[1],
-		       FI_UINT64, FI_ATOMIC_WRITE, target);
+			   loc_mr[0], gni_addr[1],
```
Why change the indentation here?

jswaro replied:
I was changing the lines here and decided I'd modify the indentation to fit the style used in other locations in this file to maintain a uniform convention.

chuckfossen commented on prov/gni/test/mr.c:
```diff
@@ -324,6 +354,56 @@ Test(mr_internal_bare, basic_init)
+
+	/* registrations must be made on aligned boundaries */
+	base = sbrk(0);
+	if ((uint64_t) base & 0xfff) {
```
Same comment as above using macro instead of hardcoded values.

jshimek commented on man/fi_gni.7.md:
```diff
@@ -277,6 +325,36 @@ address format. In order to populate the remote peer's address vector
 with this mechanism, the application must call fi_cq_readerr to get the
 source address followed by fi_av_insert on the populated err_data member.
 
+The GNI provider can use a maximum of 4K memory registrations per *node* when using scalable memory registration.
```
4K as in 4000 or 4K as in 4096?

jswaro replied:
4096.

jshimek commented on prov/gni/include/gnix_mr.h:
```diff
@@ -45,11 +45,14 @@
 #include "gnix_priv.h"
 #include "gnix_mr_cache.h"
 
+#define USER_REGISTRATION 0
```
While USER_REGISTRATION and PROV_REGISTRATION are only used in the code I would prefer it if they had the standard GNIX prefix.

jswaro replied:
They should be namespace'd appropriately. I'll fix that.

jshimek commented on prov/gni/src/gnix_rma.c:
```diff
@@ -158,17 +158,25 @@ static void __gnix_rma_copy_chained_get_data(struct gnix_fab_req *req)
 		memcpy((void *)addr,
-		       (void *) ((uint8_t *) req->int_tx_buf + GNI_READ_ALIGN),
-		       tail_len);
+			   (void *) ((uint8_t *) req->int_tx_buf + GNI_READ_ALIGN),
```
Is there a point to adjusting the white-space here it actually looks like the spacing is indented incorrectly now.

jshimek commented on prov/gni/src/gnix_rma.c:
```diff
@@ -158,17 +158,25 @@ static void __gnix_rma_copy_chained_get_data(struct gnix_fab_req *req)
+		rc = fi_close(&req->rma.loc_md->mr_fid.fid);
+		if (rc != FI_SUCCESS)
+			GNIX_FATAL(FI_LOG_DOMAIN,
+				"failed to close auto-registration, "
```
At least in githubs system the "failed" falls under the AL of FATAL. If it under the FI_LOG_DOMAIN in the regular code this is fine and the comment can be ignored

jswaro commented:
Aesthetic issues aside: convention of format differences of opinion, I've addressed most of the comments. Feel free to re-review the new commits. I'll rebase/squash at end of day.

chuckfossen commented on prov/gni/test/allocator.c:
```diff
@@ -176,18 +186,24 @@ static void open_close_allocator(enum gnix_page_size page_size,
 	cr_assert_eq(ret, FI_SUCCESS, "_gnix_mbox_allocator_create failed5.");
 	cr_expect_eq(verify_hugepages(), 2 + ALLOCD_WITH_NIC,
-		     "memory not found in /proc/self/maps.");
+			 "memory not found in /proc/self/maps.");
```
You still have an extra space after the leading tabs.

chuckfossen commented on prov/gni/test/allocator.c:
```diff
@@ -176,18 +186,24 @@ static void open_close_allocator(enum gnix_page_size page_size,
 	cr_assert_eq(ret, FI_SUCCESS, "_gnix_mbox_allocator_destroy failed.");
 	cr_expect_eq(verify_hugepages(), 1 + ALLOCD_WITH_NIC,
-		     "memory not released in /proc/self/maps.");
+			 "memory not released in /proc/self/maps.");
```
and here

jswaro commented:
bot:lanl:retest

jswaro commented:
bot:lanl:retest

jswaro commented:
bot:lanl:retest
